### ISSUE TITLE
[windows] duplicated items in v:oldfiles

### ISSUE THREAD

**Issue by skywind3000** [neovim/neovim#25033]
### Problem

Firstly I thought it was a telescope bug:

Same file, `E:\Lab\timer.py`, got three duplicated items in the telescope oldfiles

![图片](https://github.com/neovim/neovim/assets/3035071/03840964-a287-4aeb-bae2-901d1003d0ae)

Then I realized to check the `v:oldfiles` variable:

![图片](https://github.com/neovim/neovim/assets/3035071/91712e48-4d30-4af1-b32c-840423462d4b)

And I found it was caused by nvim itself, which records three different records of same file in the v:oldfiles.

### Steps to reproduce

Sometimes, I can reproduce by:

```viml
:tabe e:/lab/timer.py
:close
:tabe e:\lab/timer.py
:close
```

but sometimes I can't.

### Expected behavior

Same file have only one entry in the v:oldfiles

### Neovim version (nvim -v)

0.9.1

### Vim (not Nvim) behaves the same?

No, Vim 9.0.1839

### Operating system/version

Windows 10

### Terminal name/version

nvim-qt

### $TERM environment variable

None

### Installation

repo

**Comment by justinmk** [comment]
Inconsistent slashes on Windows are a known issue, we need a Windows dev to take a close look at this.

My [general sense](https://github.com/neovim/neovim/issues/23639#issuecomment-1548647316) is that a solution for this would look like:

1. To start with, _remove_ all `\` slashes entirely. So internally we are always dealing with `/` slashes. 
2. Then address particular cases such as `fnamemodify` where `\` are _explicitly requested_ and only _then_ output the path with `\` slashes.

Related:

- https://github.com/neovim/neovim/pull/23793
- https://github.com/neovim/neovim/issues/23639
- https://github.com/neovim/neovim/issues/15966

**Comment by sid-6581** [comment]
I was having this issue (and more) with slashes on Windows, and I stumbled across this issue while looking for possible solutions. Here are my random initial observations and thoughts:

1. Turning on shellslash breaks Telescope, and it appears to be a bug in plenary. If I remember correctly it seems to have some hardcoded dependency on backslashes on Windows.

2. I added the following autocmd as a very hacky bandaid to just replace backslashes in buffer names with forward slashes, as well as capitalizing the drive letter:

```lua
-- Normalize paths for Windows buffers
if vim.fn.has("win32") == 1 then
  vim.api.nvim_create_autocmd({ "BufAdd" }, {
    callback = function()
      local name = vim.api.nvim_buf_get_name(0)
      if name:sub(2, 2) == ":" then
        name = name:gsub("\\", "/"):gsub("^%l", string.upper)
        vim.api.nvim_buf_set_name(0, name)
      end
    end,
  })
end
```

The drive letter capitalization is needed because everyone seems to do their own normalization of file paths, and for some that includes lower casing the drive letter. Some picky LSPs like taplo that appear to just do a case sensitive string comparison of paths fail when the drive letter case doesn't match.

This autocmd works somewhat, but not in all cases. Depending on how the buffer was opened I still get backslashes. I could go through all buffers and correct them before neovim exits and writes to the shada file, but it's still a big hack and not a real solution.

3. This above approach does *not* break Telescope/plenary, so I'm hoping that if neovim changes all internal paths to `/`, that would also work.

4. For Windows there are also UNC paths, and device prefixes like `\\?\C:\` to account for. Here is a good reference: https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats

I would be willing to try to take a look at solving this problem once and for all in neovim internally, since I use neovim on both Windows and Linux extensively. I have contributed Windows functionality to vim before, and I'm looking through the neovim codebase to familiarize myself with it as well.

@justinmk are you aware of anyone else working on this? I don't want to duplicate any efforts. I agree that working internally with just `/` seems like the cleanest approach.

**Comment by justinmk** [comment]
> @justinmk are you aware of anyone else working on this? I don't want to duplicate any efforts. I agree that working internally with just `/` seems like the cleanest approach.

I linked a closed PR above. Other than that, no.

**Comment by sid-6581** [comment]
Sounds good! I'll go through the issues and PRs to make sure, then dig into it and see what I find. This path issue has been such a thorn in my side that it would be nice to get it solved cleanly, if possible. Right now all I'm seeing is sketchy workarounds in every plugin I've had problems with.

**Comment by justinmk** [comment]
Yeah this is a painful bug, fixing it will be greatly appreciated.

**Comment by sid-6581** [comment]
I've poked around and tested a couple of things, and I think this should be doable. I also noticed other places that are completely broken on Windows, like `path_fix_case` that is hardcoded to look for `/` only.

I think as part of this it makes sense to fully normalize the file path on Windows, including fixing the case of all directory names to eliminate another source of duplicate entries in various places as well as incorrect comparisons in LSPs or plugins that do case sensitive comparisons. `path_fix_case`, despite the name, only fixes the file name and not the full path, which doesn't seem to be enough on Windows.

I'll keep working on it over the weekend, but initial impressions is that it shouldn't be too bad. Nothing has broken so far (including plugins) from testing out the changes, but there's more to do.

**Comment by brglng** [comment]
Forward slash `/` does *NOT* always work on Windows. Especially, UNC paths like `\\?\C:\` does not accept forward slashes. I believe the current `vim.fs.normalize` implementation is not correct in this situation. I personally think that always using forward slashes on Windows is a better idea. However this requires plugin authors to always use `vim.fs.*` instead of manipulating paths using forward slashes by themselves.

**Comment by justinmk** [comment]
> UNC paths like `\\?\C:\` does not accept forward slashes

Provide a reference and say exactly what steps you took to confirm this. Almost all tools, except some cmd.exe builtins, should work with "/" slashes.

> I personally think that always using forward slashes on Windows is a better idea

Nonsense. Internally, "/" slashes can always be used. It ONLY matters when the paths reach specific external tools.

**Comment by brglng** [comment]
> Provide a reference and say exactly what steps you took to confirm this. Almost all tools, except some cmd.exe builtins, should work with "/" slashes.

I found that information here: https://superuser.com/questions/176388/why-does-windows-use-backslashes-for-paths-and-unix-forward-slashes .

You can also find some information from Microsoft here: https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation , which states:

> File I/O functions in the Windows API convert "/" to "\\" as part of converting the name to an NT-style name, except when using the "\\\\?\" prefix as detailed in the following sections.

> The "\\\\?\" prefix can also be used with paths constructed according to the universal naming convention (UNC). To specify such a path using UNC, use the "\\\\?\\UNC\\" prefix. For example, "\\\\?\\UNC\\server\\share", where "server" is the name of the computer and "share" is the name of the shared folder. These prefixes are not used as part of the path itself. They indicate that the path should be passed to the system with minimal modification, which means that you cannot use forward slashes to represent path separators, or a period to represent the current directory, or double dots to represent the parent directory. Because you cannot use the "\\\\?\\" prefix with a relative path, relative paths are always limited to a total of MAX_PATH characters.

I have tested using the "Run" dialog (Win+R), and found it cannot open the file explorer if using `//?/C:/` instead of `\\?\C:\`, but I have not tested the underlying API yet.

**Comment by brglng** [comment]
> Nonsense. Internally, "/" slashes can always be used. It ONLY matters when the paths reach specific external tools.

Sorry, I mistyped in my post by mistyping "backslashes" as "forward slashes". Actually I meant that using backslashes on Windows is better, since it covers more corner cases. However I agree that users can still use forward slashes and Neovim can convert them to backslashes before passing the paths to Windows API.

**Comment by ntdiary** [comment]
FYI:
https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#skip-normalization
Paths starting with `\\?\` aren not normalized, possibly for the following reasons:
> 1. To get access to paths that are normally unavailable but are legal. **A file or directory called `hidden.`**, for example, is impossible to access in any other way.
> 2. To improve performance by skipping normalization **if you've already normalized**.
> 3. On .NET Framework only, to skip the MAX_PATH check for path length to allow for paths that are greater than 259 characters. Most APIs allow this, with some exceptions.

<img height="400" alt="Image" src="https://github.com/user-attachments/assets/80cc9e9a-80d7-424c-9af7-533791c4d8fe" />

The case in point 1 also appears in #31606 (i.e. `test.\file.txt`), maybe we could treat this as a special edge case. 😅

**Comment by ntdiary** [comment]
In nightly builds, path separators have been normalized to `/` (except for URI), though there's still a case-sensitivity issue that needs a separate PR. As for edge cases like `\\?\C:\hidden.\bar.md`, we could fix them after the [root parsing behavior](https://github.com/neovim/neovim/issues/39844) is finalized. 

<img height="100" alt="Image" src="https://github.com/user-attachments/assets/7e0e9ccf-b35e-4229-a735-5fa80da538bc" />
